src/Common/Crypto/X509Certificate.h

```diff
@@ -0,0 +1,333 @@
+    {
+        using BIO_ptr = std::unique_ptr<BIO, decltype(&BIO_free)>;
+        BIO_ptr bio(BIO_new(BIO_s_mem()), BIO_free);
+
+        if (!bio)
+            throw Exception(ErrorCodes::OPENSSL_ERROR, "BIO_new failed: {}", getOpenSSLErrors());
+
+        BIO * file = BIO_new_file(path.c_str(), "r");
+
+        if (!file)
+            throw Exception(ErrorCodes::OPENSSL_ERROR, "BIO_new_file failed: {}", getOpenSSLErrors());
+
+        certificate = PEM_read_bio_X509(file, nullptr, nullptr, nullptr);
+        if (!certificate)
+            throw Exception(ErrorCodes::OPENSSL_ERROR, "PEM_read_bio_X509 failed for file {}: {}", path, getOpenSSLErrors());
+    }
+
+    static X509Certificate::List fromFile(const std::string & path)
+    {
+        using BIO_ptr = std::unique_ptr<BIO, decltype(&BIO_free)>;
+
+        BIO_ptr bio(BIO_new_file(path.c_str(), "r"), BIO_free);
+        if (!bio)
+            throw Exception(ErrorCodes::OPENSSL_ERROR, "BIO_new_file failed: {}", getOpenSSLErrors());
+
+        X509Certificate::List certs;
+
+        while (true)
+        {
+            X509 * cert = PEM_read_bio_X509(bio.get(), nullptr, nullptr, nullptr);
+
+            if (!cert)
+            {
+                auto err = ERR_peek_last_error();
+                if (err == 0)
+                    break;
+
+                /// We read at least one cert, and can't find
+                /// the beginning of a next one.
+                /// This most likely means we reached the end of the file.
+                if (!certs.empty()
+                    /// Manually unwrap ERR_GET_REASON(err) due to ossl_unused
+                    /// https://github.com/openssl/openssl/issues/16776
+                    ///
+                    /// To be fixed in OpenSSL 3.4+
+                    && ((err & ERR_SYSTEM_FLAG) == 0 && (err & ERR_REASON_MASK) == PEM_R_NO_START_LINE))
+                    /// Means we reached the end of the file.
+                    break;
+
+                throw Exception(ErrorCodes::OPENSSL_ERROR, "PEM_read_bio_X509 failed: c:{}, f:{}, e:{}", certs.size(), path, getOpenSSLErrors());
+            }
+
+            certs.emplace_back(cert);
+        }
+
+        if (certs.empty())
+            throw Exception(ErrorCodes::OPENSSL_ERROR, "No certificates found in file: {}", path);
+
+        return certs;
+    }
+
+    ~X509Certificate()
+    {
+        if (certificate)
+            X509_free(certificate);
+    }
+
+    uint64_t version() const
+    {
+        // This is defined by standards (X.509 et al) to be
+        // one less than the certificate version.
+        // So, eg. a version 3 certificate will return 2.
+        return X509_get_version(certificate) + 1;
+    }
+
+    std::string serialNumber() const
+    {
+        ASN1_INTEGER * serial = X509_get_serialNumber(certificate);
+        BIGNUM * bn = ASN1_INTEGER_to_BN(serial, nullptr);
+        char * hex = BN_bn2hex(bn);
+        std::string result(hex);
```

Comment:
The memory for `hex` and `bn` could theoretically leak if `result cannot be allocated. We should use RAII to de-alloc automatically.

Something like

```cpp
SCOPE_EXIT(
        {
            BN_free(bn);
        });
```

(like before) will do the job.